src/error.c

```diff
@@ -15,97 +15,102 @@
     ECode(WrongDoubleQuote, "User entered `“` (0x201c) as a quote instead of ASCII"), \
     ECode(Syntax, "Syntax error"), \
     ECode(Fatal, "Fatal error"), \
+    ECode(Redefinition, "Redefinition"), \
 
 typedef enum ErrorCode {
 #define ECode(e, s) e##Error
     ERROR_CODES
 #undef ECode
 } ErrorCode;
 
-typedef struct Error {
-    ErrorCode code;
-    Position pos;
-    const char *message;
-} Error;
-
-typedef struct ErrorCollector {
-    u32 errorCount;
-} ErrorCollector;
+typedef struct DiagnosticNote DiagnosticNote;
+struct DiagnosticNote {
+    const char *msg;
+    DiagnosticNote *next;
+};
 
-ErrorCollector errorCollector;
+typedef struct DiagnosticError DiagnosticError;
+struct DiagnosticError {
+    const char *msg;
+    DiagnosticNote *note;
+};
 
-b32 HasErrors() {
-    return errorCollector.errorCount > 0;
+b32 shouldPrintErrorCode() {
+#if NO_ERROR_CODES
+    return false;
+#endif
+    return FlagErrorCodes;
 }
 
-#define tempErrorBufferLen KB(4)
-char *temporaryErrorBuffer;
+void ReportError(Package *p, ErrorCode code, Position pos, const char *msg, ...) {
+    va_list args;
+    char msgBuffer[512]; // TODO: Static & Thread Local?
+    va_start(args, msg);
+    vsnprintf(msgBuffer, sizeof(msgBuffer), msg, args);
+    char errorBuffer[512];
+    int errlen = shouldPrintErrorCode() ?
+        snprintf(errorBuffer, sizeof(errorBuffer), "ERROR(%s:%u:%u, E%04d): %s\n", pos.name, pos.line, pos.column, code, msgBuffer) :
+        snprintf(errorBuffer, sizeof(errorBuffer), "ERROR(%s:%u:%u): %s\n",        pos.name, pos.line, pos.column,       msgBuffer);
+    char *errorMsg = ArenaAlloc(&p->diagnostics.arena, errlen + 1);
```

Comment:
IIRC `snprintf` returns the length the string **would have** been had it had a large enough buffer. We should probably min clamp this to protect against buffer-overrun.